### PR TITLE
[REVIEW] Add index support in `DataFrame.query`

### LINKED ISSUE #5195: [BUG] KeyError when referencing index column in DataFrame.query()
**Describe the bug**
After setting a column as index, trying to use it in DataFrame.query() expression raises KeyError.

**Steps/Code to reproduce bug**
Run this code:

```
import numpy as np
import cudf
from datetime import datetime, timedelta

t0 = datetime.strptime('2018-10-07 12:00:00', '%Y-%m-%d %H:%M:%S')
n = 5

df = cudf.DataFrame()
df['timestamp'] = np.array([(t0 + timedelta(seconds=x)) for x in range(n)])
df['value'] = np.arange(n)

start = datetime.strptime('2018-10-07 12:00:02', '%Y-%m-%d %H:%M:%S')

df.query('timestamp > @start') # This works fine

df = df.set_index('timestamp')
df.query('timestamp > @start') # KeyError on line 223 of cudf/core/column_accessor.py
```
Replacing 'timestamp' with 'index' in query() raises the same KeyError.
The same code works perfect in Pandas, using either 'timestamp' or 'index' identifiers.

**Expected behavior**
Successful DataFrame.query() execution.

**Environment overview (please complete the following information)**
 - Environment location: Google Colab
 - Method of cuDF install: conda (using rapidsai-csp-utils/colab/rapids-colab.sh)

**Environment details**
<details><summary>Click here to see environment details</summary><pre>
     
     **git***
print_env.sh: 10: [: unexpected operator
     Not inside a git repository
     
     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=18.04
     DISTRIB_CODENAME=bionic
     DISTRIB_DESCRIPTION="Ubuntu 18.04.3 LTS"
     NAME="Ubuntu"
     VERSION="18.04.3 LTS (Bionic Beaver)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 18.04.3 LTS"
     VERSION_ID="18.04"
     HOME_URL="https://www.ubuntu.com/"
     SUPPORT_URL="https://help.ubuntu.com/"
     BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
     PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
     VERSION_CODENAME=bionic
     UBUNTU_CODENAME=bionic
     Linux cdb7d4d8cbe9 4.19.104+ #1 SMP Wed Feb 19 05:26:34 PST 2020 x86_64 x86_64 x86_64 GNU/Linux
     
     ***GPU Information***
     Thu May 14 08:38:41 2020       
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 440.82       Driver Version: 418.67       CUDA Version: 10.1     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |===============================+======================+======================|
     |   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
     | N/A   36C    P0    31W / 250W |    483MiB / 16280MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     
     +-----------------------------------------------------------------------------+
     | Processes:                                                       GPU Memory |
     |  GPU       PID   Type   Process name                             Usage      |
     |=============================================================================|
     +-----------------------------------------------------------------------------+
     
     ***CPU***
     Architecture:        x86_64
     CPU op-mode(s):      32-bit, 64-bit
     Byte Order:          Little Endian
     CPU(s):              2
     On-line CPU(s) list: 0,1
     Thread(s) per core:  2
     Core(s) per socket:  1
     Socket(s):           1
     NUMA node(s):        1
     Vendor ID:           GenuineIntel
     CPU family:          6
     Model:               63
     Model name:          Intel(R) Xeon(R) CPU @ 2.30GHz
     Stepping:            0
     CPU MHz:             2300.000
     BogoMIPS:            4600.00
     Hypervisor vendor:   KVM
     Virtualization type: full
     L1d cache:           32K
     L1i cache:           32K
     L2 cache:            256K
     L3 cache:            46080K
     NUMA node0 CPU(s):   0,1
     Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt arat md_clear arch_capabilities
     
     ***CMake***
     /usr/local/bin/cmake
     cmake version 3.12.0
     
     CMake suite maintained and supported by Kitware (kitware.com/cmake).
     
     ***g++***
     /usr/bin/g++
     g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
     Copyright (C) 2017 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
     
     
     ***nvcc***
     /usr/local/cuda/bin/nvcc
     nvcc: NVIDIA (R) Cuda compiler driver
     Copyright (c) 2005-2019 NVIDIA Corporation
     Built on Sun_Jul_28_19:07:16_PDT_2019
     Cuda compilation tools, release 10.1, V10.1.243
     
     ***Python***
     /usr/local/bin/python
     Python 3.6.10
     
     ***Environment Variables***
     PATH                            : /usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/tools/node/bin:/tools/google-cloud-sdk/bin:/opt/bin
     LD_LIBRARY_PATH                 : /usr/lib64-nvidia
     NUMBAPRO_NVVM                   : 
     NUMBAPRO_LIBDEVICE              : 
     CONDA_PREFIX                    : 
     PYTHON_PATH                     : 
     
     ***conda packages***
     conda is /usr/local/bin/conda
     /usr/local/bin/conda
     # packages in environment at /usr/local:
     #
     # Name                    Version                   Build  Channel
     _libgcc_mutex             0.1                 conda_forge    conda-forge
     _openmp_mutex             4.5                       0_gnu    conda-forge
     arrow-cpp                 0.15.0           py36h090bef1_2    conda-forge
     asn1crypto                0.24.0                   py36_0  
     bokeh                     2.0.1            py36h9f0ad1d_0    conda-forge
     boost-cpp                 1.70.0               h8e57a91_2    conda-forge
     brotli                    1.0.7             he1b5a44_1001    conda-forge
     bzip2                     1.0.8                h516909a_2    conda-forge
     c-ares                    1.15.0            h516909a_1001    conda-forge
     ca-certificates           2020.4.5.1           hecc5488_0    conda-forge
     certifi                   2020.4.5.1       py36h9f0ad1d_0    conda-forge
     cffi                      1.11.5           py36h9745a5d_0  
     chardet                   3.0.4            py36h0f667ec_1  
     click                     7.1.2              pyh9f0ad1d_0    conda-forge
     cloudpickle               1.4.1                      py_0    conda-forge
     conda                     4.5.4                    py36_0  
     conda-env                 2.6.0                h36134e3_1  
     contextvars               2.4                        py_0    conda-forge
     cryptography              2.2.2            py36h14c3975_0  
     cudatoolkit               10.1.243             h6bb024c_0    nvidia
     cudf                      0.13.0                   py36_0    rapidsai/label/main
     cudnn                     7.6.0                cuda10.1_0    nvidia
     cupy                      7.4.0            py36h5c369b2_2    conda-forge
     cytoolz                   0.10.1           py36h516909a_0    conda-forge
     dask                      2.16.0                     py_0    conda-forge
     dask-core                 2.16.0                     py_0    conda-forge
     dask-cudf                 0.13.0                   py36_0    rapidsai/label/main
     distributed               2.16.0           py36h9f0ad1d_0    conda-forge
     dlpack                    0.2                  he1b5a44_1    conda-forge
     double-conversion         3.1.5                he1b5a44_2    conda-forge
     fastavro                  0.23.3           py36h8c4c3a4_0    conda-forge
     fastrlock                 0.4             py36h831f99a_1001    conda-forge
     freetype                  2.10.2               he06d7ca_0    conda-forge
     fsspec                    0.6.3                      py_0    conda-forge
     gflags                    2.2.2             he1b5a44_1002    conda-forge
     glog                      0.4.0                h49b9bf7_3    conda-forge
     grpc-cpp                  1.23.0               h18db393_0    conda-forge
     heapdict                  1.0.1                      py_0    conda-forge
     icu                       64.2                 he1b5a44_1    conda-forge
     idna                      2.6              py36h82fb2a8_1  
     immutables                0.12             py36h8c4c3a4_0    conda-forge
     jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
     jpeg                      9c                h14c3975_1001    conda-forge
     ld_impl_linux-64          2.34                 h53a641e_0    conda-forge
     libblas                   3.8.0               14_openblas    conda-forge
     libcblas                  3.8.0               14_openblas    conda-forge
     libcudf                   0.13.0               cuda10.1_0    rapidsai/label/main
     libedit                   3.1.20170329         h6b74fdf_2  
     libevent                  2.1.10               h72c5cf5_0    conda-forge
     libffi                    3.2.1             he1b5a44_1007    conda-forge
     libgcc-ng                 9.2.0                h24d8f2e_2    conda-forge
     libgfortran-ng            7.3.0                hdf63c60_5    conda-forge
     libgomp                   9.2.0                h24d8f2e_2    conda-forge
     liblapack                 3.8.0               14_openblas    conda-forge
     libllvm8                  8.0.1                hc9558a2_0    conda-forge
     libnvstrings              0.13.0               cuda10.1_0    rapidsai/label/main
     libopenblas               0.3.7                h5ec1e0e_6    conda-forge
     libpng                    1.6.37               hed695b0_1    conda-forge
     libprotobuf               3.8.0                h8b12597_0    conda-forge
     librmm                    0.13.0               cuda10.1_0    rapidsai/label/main
     libstdcxx-ng              9.2.0                hdf63c60_2    conda-forge
     libtiff                   4.1.0                hfc65ed5_0    conda-forge
     llvmlite                  0.31.0           py36hfa65bc7_1    conda-forge
     locket                    0.2.0                      py_2    conda-forge
     lz4-c                     1.8.3             he1b5a44_1001    conda-forge
     markupsafe                1.1.1            py36h8c4c3a4_1    conda-forge
     msgpack-python            1.0.0            py36hdb11119_1    conda-forge
     nccl                      2.4.6.1              cuda10.1_0    nvidia
     ncurses                   6.1               hf484d3e_1002    conda-forge
     numba                     0.48.0           py36hb3f55d8_0    conda-forge
     numpy                     1.18.4           py36h7314795_0    conda-forge
     nvstrings                 0.13.0                   py36_0    rapidsai/label/main
     olefile                   0.46                       py_0    conda-forge
     openssl                   1.1.1g               h516909a_0    conda-forge
     packaging                 20.1                       py_0    conda-forge
     pandas                    0.25.3           py36hb3f55d8_0    conda-forge
     parquet-cpp               1.5.1                         2    conda-forge
     partd                     1.1.0                      py_0    conda-forge
     pillow                    7.1.2            py36h8328e55_0    conda-forge
     pip                       20.1               pyh9f0ad1d_0    conda-forge
     psutil                    5.7.0            py36h8c4c3a4_1    conda-forge
     pyarrow                   0.15.0           py36h8b68381_1    conda-forge
     pycosat                   0.6.3            py36h0a5515d_0  
     pycparser                 2.18             py36hf9f622e_1  
     pyopenssl                 18.0.0                   py36_0  
     pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
     pysocks                   1.6.8                    py36_0  
     python                    3.6.10          h8356626_1011_cpython    conda-forge
     python-dateutil           2.8.1                      py_0    conda-forge
     python_abi                3.6                     1_cp36m    conda-forge
     pytz                      2020.1             pyh9f0ad1d_0    conda-forge
     pyyaml                    5.1.2            py36h516909a_0    conda-forge
     re2                       2020.05.01           he1b5a44_0    conda-forge
     readline                  8.0                  hf8c457e_0    conda-forge
     requests                  2.18.4           py36he2e5f8d_1  
     rmm                       0.13.0                   py36_0    rapidsai/label/main
     ruamel_yaml               0.15.37          py36h14c3975_2  
     setuptools                46.3.0           py36h9f0ad1d_0    conda-forge
     six                       1.11.0           py36h372c433_1  
     snappy                    1.1.8                he1b5a44_1    conda-forge
     sortedcontainers          2.1.0                      py_0    conda-forge
     sqlite                    3.30.1               hcee41ef_0    conda-forge
     tblib                     1.6.0                      py_0    conda-forge
     thrift-cpp                0.12.0            hf3afdfd_1004    conda-forge
     tk                        8.6.10               hed695b0_0    conda-forge
     toolz                     0.10.0                     py_0    conda-forge
     tornado                   6.0.4            py36h8c4c3a4_1    conda-forge
     typing_extensions         3.7.4.2                    py_0    conda-forge
     uriparser                 0.9.3                he1b5a44_1    conda-forge
     urllib3                   1.22             py36hbe7ace6_0  
     wheel                     0.34.2                     py_1    conda-forge
     xz                        5.2.5                h516909a_0    conda-forge
     yaml                      0.1.7                had09818_2  
     zict                      2.0.0                      py_0    conda-forge
     zlib                      1.2.11            h516909a_1006    conda-forge
     zstd                      1.4.3                h3b9ef0a_0    conda-forge
     
</pre></details>

Tested on both 0.13 stable and 0.14 nightly.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
This adds `index` support to `DataFrame.query`,  where earlier index were not included as part of query and index names in query would have resulted in `KeyError`, now it works as expected.

This also adds support of `index` keyword in query, this would fail in case `cudf.MultiIndex`.

close #5195

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/utils/applyutils.py [hunk 1]
```diff
@@ -111,12 +111,27 @@ def apply_chunks(
 
 
 def make_aggregate_nullmask(df, columns=None, op="and"):
+    from cudf import MultiIndex
```
Comment (kkraus14) [inline python/cudf/cudf/utils/applyutils.py:114]: Don't import in the function, just use `cudf.MultiIndex` below

File: python/cudf/cudf/utils/queryutils.py [hunk 2]
```diff
@@ -194,6 +194,8 @@ def query_execute(df, expr, callenv):
         Contains keys 'local_dict', 'locals' and 'globals' which are all dict.
         They represent the arg, local and global dictionaries of the caller.
     """
+    from cudf import MultiIndex
```
Comment (kkraus14) [inline python/cudf/cudf/utils/queryutils.py:197]: Don't import in the function

File: python/cudf/cudf/utils/queryutils.py [hunk 3]
```diff
@@ -215,7 +217,21 @@ def query_execute(df, expr, callenv):
...
+        try:
+            return df[col]._column.data_array_view
+        except KeyError:
+            if (
+                col == "index"
+                and col not in df.index._data
+                and not isinstance(df.index, MultiIndex)
+            ):
+                return df.index._data.columns[0].data_array_view
+            return df.index._data[col].data_array_view
```
Comment (kkraus14) [inline python/cudf/cudf/utils/queryutils.py:231]: We shouldn't need to redefine this in each function.

File: python/cudf/cudf/utils/applyutils.py [hunk 4]
```diff
@@ -111,12 +111,27 @@ def apply_chunks(
...
     out_mask = None
-    for k in columns or df.columns:
-        if not df[k].nullable:
-            continue
+    for k in columns or df._data:
+        try:
+            if not df[k].nullable:
+                continue
+            nullmask = df[k].nullmask
+        except KeyError:
```
Comment (kkraus14) [inline python/cudf/cudf/utils/applyutils.py:122]: I typically prefer to check if something exists as opposed to just catching the exception. But would be curious for others thoughts as well.

File: python/cudf/cudf/utils/queryutils.py [hunk 5]
```diff
@@ -215,7 +217,21 @@ def query_execute(df, expr, callenv):
             envargs.append(val)
     columns = compiled["colnames"]
     # prepare col args
-    colarrays = [df[col]._column.data_array_view for col in columns]
+
+    def extract_colarrays(col):
+        try:
+            return df[col]._column.data_array_view
+        except KeyError:
```
Comment (kkraus14) [inline python/cudf/cudf/utils/queryutils.py:224]: Similar here.

File: python/cudf/cudf/utils/applyutils.py (hunk 1 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/utils/applyutils.py:114]: cudf isn't imported in the file, so I might have to import cudf before that.

File: python/cudf/cudf/utils/queryutils.py (hunk 2 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/utils/queryutils.py:197]: cudf isn't imported in the file, so I might have to import cudf before that. So not sure if that's a better option.

Comment (kkraus14) [reply python/cudf/cudf/utils/queryutils.py:197]: Importing cudf is fine. In general we've seen that if these are used in tight loops they get really expensive versus just the module loading once.

File: python/cudf/cudf/utils/applyutils.py (hunk 4 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/utils/applyutils.py]: I tried an example with `try and except` and with `search`, overall `try and except` looks better compared to other option
```
In [46]: import cudf                                                                                                                                                                                                             

In [47]: df = cudf.DataFrame({"a":[1], "b":[2], "c":[1], "d":[3], "e":[4]})                                                                                                                                                      

In [48]: df = df.set_index('a')                                                                                                                                                                                                  

In [49]: def sear(df, col): 
    ...:     if col in df.index._data: 
    ...:         return df.index._data[col] 
    ...:     else: 
    ...:         return df._data[col] 
    ...:                                                                                                                                                                                                                         

In [50]: def try_ex(df, col): 
    ...:     try: 
    ...:         return df._data[col] 
    ...:     except KeyError: 
    ...:         return df.index._data[col] 
    ...:          
    ...:                                                                                                                                                                                                                         

In [51]: %timeit try_ex(df, 'a')                                                                                                                                                                                                 
895 ns ± 0.569 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [52]: %timeit sear(df, 'a')                                                                                                                                                                                                   
868 ns ± 1.14 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [53]: %timeit try_ex(df, 'e')                                                                                                                                                                                                 
307 ns ± 0.431 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [54]: %timeit sear(df, 'e')                                                                                                                                                                                                   
971 ns ± 0.694 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

```

Comment (kkraus14) [reply python/cudf/cudf/utils/applyutils.py]: You have the logic backwards 😄 It searches in `df._data` before searching in `df.index._data`.

Comment (rgsl888prabhu) [reply python/cudf/cudf/utils/applyutils.py]: For try_except I left it after as I had to take care of situation `index` keyword. And didn't want to search again in `df._data` for it.

Comment (rgsl888prabhu) [reply python/cudf/cudf/utils/applyutils.py]: With switched logic 
```
In [56]: def sear(df, col): 
    ...:     if col in df._data: 
    ...:         return df._data[col] 
    ...:     else: 
    ...:         return df.index._data[col] 
    ...:          
    ...:                                                                                                                                                                                                                         

In [57]: %timeit sear(df, 'a')                                                                                                                                                                                                   
975 ns ± 0.448 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [58]: %timeit sear(df, 'e')                                                                                                                                                                                                   
641 ns ± 1.16 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

```

Comment (kkraus14) [reply python/cudf/cudf/utils/applyutils.py]: Seems like try except is more performant, lets go with that for now 😄

**Pull request merged**